web3.js/src/connection.ts

```diff
@@ -3023,104 +3023,168 @@ export class Connection {
 
     assert(decodedSignature.length === 64, 'signature has invalid length');
 
-    const subscriptionCommitment = commitment || this.commitment;
-    let timeoutId;
-    let subscriptionId;
-    let done = false;
+    if (
+      Object.prototype.hasOwnProperty.call(strategy, 'lastValidBlockHeight')
+    ) {
+      return await this.confirmTransactionUsingBlockheightExceedenceStrategy({
+        commitment: commitment || this.commitment,
+        strategy: strategy as BlockheightBasedTransactionConfirmationStrategy,
+        signature: rawSignature,
+      });
+    } else {
+      return await this.confirmTransactionUsingLegacyTimeoutStrategy({
+        commitment: commitment || this.commitment,
+        signature: rawSignature,
+      });
+    }
+  }
 
+  private getTransactionConfirmationPromise({
+    commitment,
+    signature,
+  }: {
+    commitment?: Commitment;
+    signature: string;
+  }): [
+    Promise<{
+      __type: TransactionStatus.PROCESSED;
+      response: RpcResponseAndContext<SignatureResult>;
+    }>,
+    () => void,
+  ] {
+    let subscriptionId: number | undefined;
     const confirmationPromise = new Promise<{
       __type: TransactionStatus.PROCESSED;
       response: RpcResponseAndContext<SignatureResult>;
     }>((resolve, reject) => {
       try {
         subscriptionId = this.onSignature(
-          rawSignature,
+          signature,
           (result: SignatureResult, context: Context) => {
             subscriptionId = undefined;
             const response = {
               context,
               value: result,
             };
-            done = true;
             resolve({__type: TransactionStatus.PROCESSED, response});
           },
-          subscriptionCommitment,
+          commitment,
         );
       } catch (err) {
         reject(err);
       }
     });
+    const cancel = () => {
+      if (subscriptionId == null) {
+        return;
+      }
+      this.removeSignatureListener(subscriptionId);
+      subscriptionId = undefined;
+    };
+    return [confirmationPromise, cancel];
+  }
 
-    const expiryPromise = new Promise<
-      | {__type: TransactionStatus.BLOCKHEIGHT_EXCEEDED}
-      | {__type: TransactionStatus.TIMED_OUT; timeoutMs: number}
-    >(resolve => {
-      if (typeof strategy === 'string') {
-        let timeoutMs = this._confirmTransactionInitialTimeout || 60 * 1000;
-        switch (subscriptionCommitment) {
-          case 'processed':
-          case 'recent':
-          case 'single':
-          case 'confirmed':
-          case 'singleGossip': {
-            timeoutMs = this._confirmTransactionInitialTimeout || 30 * 1000;
-            break;
-          }
-          // exhaust enums to ensure full coverage
-          case 'finalized':
-          case 'max':
-          case 'root':
+  private async confirmTransactionUsingBlockheightExceedenceStrategy({
```

Comment:
nit: block height is two words: BlockHeight
nit: Exceedence doesn't appear to be a word, maybe stick with Exceeded?

Reply:
But… but… it is? https://www.merriam-webster.com/dictionary/exceedance

Reply:
I spelled it wrong, tbf.